### ISSUE TITLE
Ability to define Contracts in modules and reuse them in other Contracts or Forms

### ISSUE THREAD

**Issue by dhampik** [apotonick/reform#184]
I was reading this disscussion https://github.com/apotonick/reform/issues/106 and thought that it would actually be great to reuse as modules not only Forms, but Contracts as well.

So how about moving module functionality to Contracts (suppose Forms are inherited from them, so they will also get this).

If we do so, we will be able to do the following:
1. Decouple validations from models into Contracts
2. Use those Contracts for validations for data which comes from rake tasks (like import) or external sources. Note, that using forms for that kind of stuff is not obvious at all.
3. Reuse those Contracts in Forms

So, basically, I propose the feature described here https://github.com/apotonick/reform#forms-in-modules to be implemented for Contracts.

@apotonick, what do you think about it?

**Comment by apotonick** [comment]
That totally makes sense, but that should work anyway without changing code?!?!

**Comment by dhampik** [comment]
Looks like you are right.
But probably it would be more obivous to move lib/reform/form/module.rb to lib/reform/contract/module.rb ?

**Comment by dhampik** [comment]
Since it's a breaking change probably it would be better to do it with next major release...

**Comment by dhampik** [comment]
Just in case anyone interested, I implemented sort of the following structure:
`app/contracts`
~~~
# user_contract.rb
class UserContract < Reform::Contract
  include UserValidationModule
end
~~~
`app/forms`
~~~
# log_in_form.rb
class LogInForm < Reform::Form

  property :remember_me, empty: true

  include UserValidationModule
end
~~~
`app/validations`
~~~
# user_validation_module.rb
module UserValidationModule
  include Reform::Form::Module

  model :user

  property :username
  property :password

  validates :username, presence: true, uniqueness: true
  validates :password, presence: true
end
~~~

Not sure how good or bad this approach is, so I'm open for feedback!

**Comment by apotonick** [comment]
That's exactly how it is designed to work, @dhampik.

Can I ask why you use forms _and_ contracts in parallel? I originally invented `Contract` to allow using people a separate deserialisation/population mechanism (e.g. Roar) and then run Reform on top of that. But then I slowly figured that I like the way Reform does it, I am not using contracts anymore.

**Comment by dhampik** [comment]
Well, yes, Contracts are exactly for what you say "to allow using people a separate deserialisation/population mechanism (e.g. Roar)"
The reason why I use them in parallel is because Forms are for representation (at least it's named like that), so I use Forms for forms representation and validation in views and Contracts for other stuff, like Roar.
So, main 2 reasons are:
1. Forms in rails have additional stuff included (which obviosly not needed for api-related logic, like for Roar)
2. The naming thing. It seems odd to create an object named "Form" to validate data which comes from API. Object named "Contract" fits better.

But that's just my personal opinion.

**Comment by dhampik** [comment]
For anyone found this issue and trying to use the structure described above. Suprisingly it does not work in production.

**Comment by apotonick** [comment]
I reopen this ticket as this sounds like a BUG. @dhampik Can I ask you to turn on `cache_classes` in development and see if that's the problem? Also, a minimal Rails app provoking this problem would help.

Thanks! :beers:

**Comment by dhampik** [comment]
https://github.com/dhampik/reform-issue-184-demo
running ```bin/rails s``` works fine, but if you run ```RAILS_ENV=production bin/rails s``` then you'll get an error right away. Yes, probably the issue is in the ```cache_classes``` option

**Comment by johnmeehan** [comment]
I think this is there is a spelling mistake in.
`reform-issue-184-demo/app/forms/log_in_form.rb`
on line 1.
It should be `class` not `ass`.

**Comment by fran-worley** [comment]
Any update @dhampik if this is a bug in reform it would be great to track it down. Otherwise if it's been resolved can we close?

**Comment by dhampik** [comment]
@johnmeehan you're right, funny typo that is :smile: 
@fran-worley I fixed the typo, that @johnmeehan pointed out and still I get an error
~~~
RAILS_ENV=production bin/rails s
...
Exiting
/home/dhampik/.rvm/gems/ruby-2.3.1/gems/reform-1.2.5/lib/reform/form/module.rb:15:in `block in included': undefined method `model' for UserContract:Class (NoMethodError)
...
~~~
So, seems the issue is still there.

**Comment by fran-worley** [comment]
@dhampik you're going to have to bump your reform version... At least to 2.1.

**Comment by apotonick** [comment]
Yes, that would help, 1.2 is a bit very old!

**Comment by dhampik** [comment]
@fran-worley @apotonick Well, I have updated reform gem version to 2.2.1 and added reform-rails gem.
Still have an error, though in a different file right now `gems/declarative-0.0.8/lib/declarative/heritage.rb:18`

Not sure if public API of reform has changed since 1.2 (very likely, cause it has a major version change), maybe my code is obsolete with the latest version of a gem now. And unfortunately I'm not working on a project where the described approach was about to be used, so I can hardly recollect what was the problem. Maybe the approach that I wanted to use is not applicable at all anymore. Maybe you can point out what is outdated in that demo repository https://github.com/dhampik/reform-issue-184-demo so I could fix that (have not found anything about Contracts/Forms in changelog.md though).

**Comment by apotonick** [comment]
We will remove `Contract` in Reform 3.0, as it simply doesn't fit into the architecture anymore. If a different population/deserialization mechanism is needed, you can use a separate representer (did you actually know that?). In my quest to providing the best docs in the world, I will document this soon on http://TRB.to so I think it's OK to close this.
